R/install_reqs.R

```diff
@@ -0,0 +1,211 @@
+  
+  if(length(elem) > 0){
+    for(i in elem){
+      install_needed = is_versioned_install_needed(i, installed, verbose)
+      if(install_needed){
+        if(!is.na(installed[i])) v = vmess(sprintf(BAD_VERSION,package,installed[i]),verbose)
+        
+        available_versions = get_available_versions(package)
+        available_compatibility = vapply(available_versions,
+                                         check_version,
+                                         TRUE,
+                                         target = version,
+                                         comp = comp)
+        if(!any(available_compatibility)){
+          failures = failures + 1
+          v = vmess(sprintf(NONE_EXISTS,package),verbose)
+        } else {
+          latest_compatible = names(which.max(which(available_compatibility)))
+          v = vmess(sprintf(INSTALL,package,
+                            sprintf(INSTALL_VERSION,latest_compatible)),verbose)
+          if(!dryrun){
+            tryCatch(devtools::install_version(package,
+                                               version=latest_compatible,
+                                               repos=repo,
+                                               quiet=TRUE),
+                     error=function(x){
+                       failures <<- failures + 1
+                       message(sprintf(OTHER_FAIL,i))
+                     })
+          }
+        }
+      }
+    }
+  }
+  return(failures)
+}
+
+install_reqs = function(reqs, dryrun, verbose = dryrun,
+                        repo = options()$repo[1], ...){
+  #' @param reqs results of process_requirements_file
+  #' @param dryrun if TRUE, no packages will be installed, but you can see what would have happened
+  #' @param verbose should the function be explicit about activities
+  #' @param repo what CRAN repository should be used?
+  #' @param ... values to pass to get_installed
+  #' @return data.frame of installed packages and their versions
+  
+  INSTALL = 'Installing %s %s'
+  INSTALL_VERSION = 'at version %s'
+  NOT_INSTALLED = 'NOTE: %s is not currently installed.'
+  BAD_VERSION = 'NOTE: %s is current installed at version %s which is not sufficient.'
+  NONE_EXISTS = 'ERROR: No version exists for %s which meets requirements.'
+  OTHER_FAIL = 'ERROR: Requirement %s could not be satisified for some reason.'
+  
+  installed = get_installed(...)
+  
+  failures = 0
+  
+  # Install git
+  failures = failures + 
+    install_special_req(reqs$git, '^git\\+', devtools::install_git, verbose) +
+    install_special_req(reqs$svn, '^svn\\+', devtools::install_svn, verbose) + 
+    install_special_req(reqs$bioc, '^bioc\\+', devtools::install_bioc, verbose) + 
+    install_special_req(reqs$url, '', devtools::install_url, verbose) + 
+    install_special_req(reqs$local, '', install_local, verbose) + 
+    install_unversioned(reqs$unversioned, verbose) + 
+    install_versioned(reqs$versioned, verbose)
+  
+  # if(length(reqs$svn) > 0){
```

Comment:
Is this block of commented out code stored in the version control history?

i.e. can we delete and go to version history if we need it back?

Reply:
Realized this is a really dumb question since I commented on code that’s currently committed. 

I’m going to delete the block comment, merge, and resolve conflicts with the compare version PR.